src/System.Net.Http/tests/UnitTests/StreamToStreamCopyTest.cs

```diff
@@ -0,0 +1,351 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Http.Tests
+{
+    public class StreamToStreamCopyTest
+    {
+        [Theory]
+        [MemberData(nameof(TwoBooleansWithAdditionalArg), new object[] { new object[] { 256, 8192 } })]
+        public async Task MemoryStream_To_MemoryStream(bool sourceIsExposable, bool disposeSource, int inputSize)
+        {
+            byte[] input = CreateByteArray(inputSize);
+            MemoryStream source = CreateSourceMemoryStream(sourceIsExposable, input);
+            var destination = new MemoryStream();
+
+            await StreamToStreamCopy.CopyAsync(source, destination, 4096, disposeSource);
+
+            Assert.NotEqual(disposeSource, source.CanRead);
+            if (!disposeSource)
+            {
+                Assert.Equal(input.Length, source.Position);
+            }
+
+            Assert.Equal(input, destination.ToArray());
+            Assert.Equal(input.Length, destination.Position);
+            Assert.Equal(input.Length, destination.Length);
```

Comment:
> Because destination.Length is >= in the general case, not necessarily equal.

No, the length should match the input length in these tests.

> So this feels like it's depending a bit too much on MemoryStream's implementation

How so?  It's not checking Capacity, it's checking Length, which isn't impacted by the internal workings of MemoryStream's growth algorithm.